### PR TITLE
Add operation id and network tags to Couchbase

### PR DESCRIPTION
**Pull request by randomanderson**
This pull request adds `operation_id` and network tags to Couchbase requests as specified [here](https://github.com/couchbaselabs/sdk-rfcs/blob/master/rfc/0035-rto.md#tags).  The main implementations differences from that RFC are:

* I did not implement `peer.service` because that would alter our current service naming
* I used the OpenTracing network tags instead of `peer.address`.

Version 2.6.0 of Couchbase is the first version where `operation_id` is specified on most requests so that is the base version instrumented.

### CODE REVIEW

File: dd-java-agent/instrumentation/couchbase-2.6/src/main/java/datadog/trace/instrumentation/couchbase/client/CouchbaseNetworkInstrumentation.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
+    // encode(ChannelHandlerContext ctx, REQUEST msg, List<Object> out)
+    return singletonMap(
+        isMethod()
+            .and(takesArguments(3))
+            .and(
+                takesArgument(
+                    0, named("com.couchbase.client.deps.io.netty.channel.ChannelHandlerContext")))
+            .and(takesArgument(2, named("java.util.List")))
+            .and(named("encode")),
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/couchbase-2.6/src/main/java/datadog/trace/instrumentation/couchbase/client/CouchbaseNetworkInstrumentation.java]: I generally like the method name earlier on in the evaluation.

File: dd-java-agent/instrumentation/couchbase-2.0/couchbase-2.0.gradle [hunk 2]
```diff
@@ -43,13 +43,21 @@ muzzle {
   }
 }
 
+configurations {
+  testArtifacts
+}
+
+artifacts {
+  testArtifacts testJar
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/couchbase-2.0/couchbase-2.0.gradle]: Add a comment that this is so the 2.6 tests can share classes.

File: dd-java-agent/instrumentation/couchbase-2.6/src/main/java/datadog/trace/instrumentation/couchbase/client/CouchbaseNetworkInstrumentation.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+          Tags.PEER_PORT.set(span, Integer.valueOf(remoteSocket.substring(splitIndex + 1)));
+        }
+
+        span.setTag("local.address", localSocket);
+      }
+    }
+
+    // 2.6.0 and above
+    public static void muzzleCheck(final JsonCryptoTranscoder transcoder) {
+      transcoder.documentType();
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/couchbase-2.6/src/main/java/datadog/trace/instrumentation/couchbase/client/CouchbaseNetworkInstrumentation.java]: Why is this needed in both instrumentation classes?  I usually prefer adding this only if there is misalignment.  What is the limiting factor that makes this integration only work with 2.6 instead of 2.5.x?

Comment (randomanderson) [reply dd-java-agent/instrumentation/couchbase-2.6/src/main/java/datadog/trace/instrumentation/couchbase/client/CouchbaseNetworkInstrumentation.java]: The network instrumentation relies on the core instrumentation to add the span to the `ContextStore`.  If the core instrumentation is turned off because the user is running pre-2.6, the network instrumentation will be a noop every time.  Might as well turn off the network instrumentation too.

---
`operationId()` was added to `CouchbaseRequest` in 2.6 .  Before that, for each message type, there was a different way of extracting the id (including parsing JSON strings).  Also, different versions have different message types so it would involve creating an instrumentation for each of couchbase-2.1, couchbase-2.2, etc.

Comment (tylerbenson) [reply dd-java-agent/instrumentation/couchbase-2.6/src/main/java/datadog/trace/instrumentation/couchbase/client/CouchbaseNetworkInstrumentation.java]: in that case this only needs to be added here since core advice will get muzzled automatically because of the use of `operationId()`. (IE, remove the check from the core instrumentation class.

File: dd-java-agent/instrumentation/couchbase-2.6/src/main/java/datadog/trace/instrumentation/couchbase/client/CouchbaseCoreInstrumentation.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+
+          if (request.operationId() != null) {
+            span.setTag("couchbase.operation_id", request.operationId());
+          }
+        }
+      }
+    }
+
+    // 2.6.0 and above
+    public static void muzzleCheck(final JsonCryptoTranscoder transcoder) {
```
Comment (tylerbenson) [inline dd-java-agent/instrumentation/couchbase-2.6/src/main/java/datadog/trace/instrumentation/couchbase/client/CouchbaseCoreInstrumentation.java]: I think this method can be removed.
